openmdao/components/tests/test_external_code_comp.py

```diff
@@ -508,6 +508,196 @@ def test_normal(self):
+                # files exist before execution and that the output files exist after.
+                self.options['external_input_files'] = [self.input_file, ]
+                self.options['external_output_files'] = [self.output_file, ]
+
+                self.options['command'] = [
+                    'python', 'extcode_node.py', self.input_file, self.output_file
+                ]
+
+            def apply_nonlinear(self, inputs, outputs, residuals):
+                with open(self.input_file, 'w') as input_file:
+                    n_in = self.options['n_in']
+                    input_file.write('{}\n'.format(n_in))
+                    for i_conn in range(n_in):
+                        input_file.write('{}\n'.format(inputs['I_in:{}'.format(i_conn)][0] ))
+                    n_out = self.options['n_out']
+                    input_file.write('{}\n'.format(n_out))
+                    for i_conn in range(n_out):
+                        input_file.write('{}\n'.format(inputs['I_out:{}'.format(i_conn)][0] ))
+
+                # the parent compute function actually runs the external code
+                super(NodeExternalCodeComp, self).apply_nonlinear(inputs, outputs, residuals)
+
+                # parse the output file from the external code and set the value of I
+                with open(self.output_file, 'r') as output_file:
+                    resid_V = float(output_file.read())
+
+                residuals['V'] = resid_V
+
+        class Circuit(Group):
+
+            def setup(self):
+                self.add_subsystem('n1', Node(n_in=1, n_out=2), promotes_inputs=[('I_in:0', 'I_in')])
+                self.add_subsystem('n2', NodeExternalCodeComp())  # leaving defaults
+
+                self.add_subsystem('R1', Resistor(R=100.), promotes_inputs=[('V_out', 'Vg')])
+                self.add_subsystem('R2', ResistorExternalCodeComp(R=10000.))
+                self.add_subsystem('D1', Diode(), promotes_inputs=[('V_out', 'Vg')])
+
+                self.connect('n1.V', ['R1.V_in', 'R2.V_in'])
+                self.connect('R1.I', 'n1.I_out:0')
+                self.connect('R2.I', 'n1.I_out:1')
+
+                self.connect('n2.V', ['R2.V_out', 'D1.V_in'])
+                self.connect('R2.I', 'n2.I_in:0')
+                self.connect('D1.I', 'n2.I_out:0')
+
+                self.nonlinear_solver = NewtonSolver()
+                self.nonlinear_solver.options['iprint'] = 2
+                self.nonlinear_solver.options['maxiter'] = 20
+                self.linear_solver = DirectSolver()
+
+        p = Problem()
+        model = p.model
+
+        model.add_subsystem('ground', IndepVarComp('V', 0., units='V'))
+        model.add_subsystem('source', IndepVarComp('I', 0.1, units='A'))
+        model.add_subsystem('circuit', Circuit())
+
+        model.connect('source.I', 'circuit.I_in')
+        model.connect('ground.V', 'circuit.Vg')
+
+        testlogger = TestLogger()
```

Comment:
Not sure we want TestLogger in a feature test.

Reply:
Done